util/types/field_type.go

```diff
@@ -127,7 +127,7 @@ func (ft *FieldType) CompactStr() string {
 			es = append(es, e)
 		}
 		suffix = fmt.Sprintf("('%s')", strings.Join(es, "','"))
-	case mysql.TypeTimestamp, mysql.TypeDatetime, mysql.TypeDate:
+	case mysql.TypeTimestamp, mysql.TypeDatetime, mysql.TypeDate, mysql.TypeDuration:
```

Comment:
Without this change,  'd time unique' create a duration field that it's Fsp is -1(UnspecifiedLength).
 `ft.Decimal != UnspecifiedLength`  prevent 'show create table x' print length of field. 
Applied this PR, the default fsp of duration is 0 when length of duration field is not set。So, duration should be printed length only with length > 0 as other time types do.